### ISSUE TITLE
Provide thread safe way to modify custom SimpleCache evictor

### ISSUE THREAD

**Issue by ojw28** [google/ExoPlayer#8628]
Filed in response to: https://github.com/google/ExoPlayer/pull/8589.

Most custom player components are used by ExoPlayer on the playback thread. To modify such components in a thread safe way, we suggest that developers should use ExoPlayer's message passing functionality (`ExoPlayer.createMessage`) to deliver changes on the playback thread.

For a custom `CacheEvictor` injected into `SimpleCache`, this approach doesn't work. `SimpleCache` is accessed on multiple threads, so the important thing is which mutex you hold, not which thread you're calling on.

The change suggested in the pull request was not thread safe as proposed, because it allowed non-thread-safe access and modification of `leastRecentlyUsed`. Synchronizing all of the evictor's public methods doesn't work either, because this approach will cause deadlock problems where one thread holds the the `SimpleCache` mutex, another holds the evictor's mutex, with each needing to acquire the other.

To modify a `SimpleCache` evictor, the important thing is that the caller holds the `SimpleCache` mutex, so technically the change in the pull request is safe provided the new method is always accessed like:
```
synchronized (simpleCache) {
  customEvictor.changeMaxBytes(newMaxBytes);
}
```
It's error-prone to rely on this, however, since this would break if someone (not unreasonably!) were to change `SimpleCache` to use an internal mutex in some future release of ExoPlayer.

I think ideally, `SimpleCache` would have a method to deliver a message through to the evictor whilst holding the correct mutex. So the application code could deliver an update safely using:
```
simpleCache.updateEvictor(messageType, payload);
```

This seems quite specific, but I think it's quite a specific case. `SimpleCache` doesn't have any other potentially-custom components that would reasonably require delivering messages to. The specificity also makes the API more obvious than what we'd be able to achieve with a general solution.

At the other end, there are really general solutions, such as `SimpleCache` letting you execute arbitrary code whilst holding the correct mutex. I think this would be quite hard for an application developer to discover and work out how to use correctly, however.

@tonihei - Any thoughts? Asking because IIRC you did most of the `PlayerMessage` work. Does my specific `updateEvictor` proposal, and a new `CacheEvictor.handleMessage(int, Object)` with a default no-op implementation, sound reasonable? Feel free to assign back after providing input :).

**Comment by tonihei** [comment]
>  Synchronizing all of the evictor's public methods doesn't work either, because this approach will cause deadlock problems where one thread holds the the SimpleCache mutex, another holds the evictor's mutex, with each needing to acquire the other.

I'n not sure I understand how this should pose a problem. Deadlocks between two mutexes are only a problem if two threads try to hold both mutexes together at the same time, but try to acquire them in a different order. The SimpleCache mutex and the evictor mutex are only acquired together by the code paths entering through the public SimpleCache methods. So as long as we always acquire the SimpleCache mutex first before trying to acquire the evictor mutex, no deadlock can occur. As we use the class mutex on every public SimpleCache method, this condition must always be true as far as I can see. All other code (such as the proposed `changeMaxBytes` method) that only acquires one of the mutexes is completely unaffected. 

Doesn't this imply that the code proposed in #8589 can be made completely thread-safe simply by synchronizing on all public methods of the class (including `changeMaxBytes`)?

**Comment by ojw28** [comment]
> All other code (such as the proposed changeMaxBytes method) that only acquires one of the mutexes is completely unaffected. Doesn't this imply that the code proposed in #8589 can be made completely thread-safe simply by synchronizing on all public methods of the class (including changeMaxBytes)?

The problem is that `changeMaxBytes` might acquire the `SimpleCache` mutex. It calls `evictCache`, which then calls `cache.removeSpan`, which acquires it. So synchronizing all public methods on the evictor ends up creating a code path that acquires the evictor mutex and then the `SimpleCache` mutex, where-as calls that go through the cache would acquire them in the opposite order.

**Comment by tonihei** [comment]
> The problem is that changeMaxBytes might acquire the SimpleCache mutex. It calls evictCache, which then calls cache.removeSpan, which acquires it.

Aha! So that's where the potential deadlock may come from. I see the problem now. 

The only other proposal I have would be to improve the class structure / locking structure to avoid this issue. 
  - For example by not having code that re-enters `Cache` from `CacheEvictor` (e.g. by returning `CacheSpans` to evict from `onStartFile` and similar).
  - Another option would be to set a lock object on CacheEvictor that has to be used and document that all access must be guarded by this lock. When initializing the cache, we can then set this lock to be the same lock that also guards the cache.
  - To avoid potential deadlocks, I think it's also generally advised to not call methods from another class while holding a lock. Not sure this is easily done though because releasing the lock just before calling the cache evictor may have hard to reason about concurrency edge cases.

**Comment by ojw28** [comment]
Yeah, it would be nice if `CacheEvictor` didn't call back into the cache. I'm not sure we can mandate that though. It would be avoidable for `LeastRecentlyUsedCacheEvictor`, if we returned the cache spans to evict from all of the `onX` methods as you suggest. We could change those methods to not have the `cache` argument either, which would reduce the possibility of custom evictor implementations calling back into the cache simply by making it less easy to access.

However, I think there are probably legitimate cases where `CacheEvictor` implementations may need to call back into the cache, if only to query some of the `getX` methods. Even if we were to not pass the `cache` as an argument into the `onXXX` methods, developers implementing such use cases would find a way to pass it in regardless via their own setter :).

Your lock suggestion is interesting, but as is the case with my original proposal, is not very enforceable. Given the lack of an enforceable option, I think I'll proceed with the `handleMessage` solution. Thanks for the input!

**Comment by tonihei** [comment]
> Your lock suggestion is interesting, but as is the case with my original proposal, is not very enforceable. 

You can enforce it easily by annotating the methods with `@GuardedBy("getLock()")` where `getLock()` is the new method returning the lock that needs to be used for access.
